core/vm/contracts.go

```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
 	}
 	return false32Byte, nil
 }
+
+type blake2F struct{}
+
+func (c *blake2F) RequiredGas(input []byte) uint64 {
+	return 1 //TODO: implement
+}
+
+func (c *blake2F) Run(input []byte) ([]byte, error) {
```

Comment:
> Why is all the padding needed? The caller sends along 320 bytes, but the method totally ignores bytes 192 - 216 , 256 - 286, 288 - 316 ?

@holiman I thought we always have to use 256-bit boxes for individual arguments to `staticcall`. 🤔 

I was experimenting with `abi.encodePacked` but I've run into some problems with `staticcall` and memory handling. Here is what I do:
```
    function F(bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f, uint32 rounds) public view returns (bytes32[2] memory) {
      bytes32[2] memory output;

      bytes memory args = abi.encodePacked(h[0], h[1], m[0], m[1], m[2], m[3], t[0], t[1], f, rounds);

      assembly {
            if iszero(staticcall(not(0), 0x09, args, 0xd5, output, 0x40)) {
                revert(0, 0)
            }
      }
      
      return output;
    }

    function callF() public view returns (bytes32[2] memory) {
      bytes32[2] memory h;
      h[0] = hex"6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2"; 
      h[1] = hex"510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179";
  
      bytes32[4] memory m;
      m[0] = hex"278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad";
      m[1] = hex"0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da";
      m[2] = hex"0000000000000000000000000000000000000000000000000000000000000000";
      m[3] = hex"0000000000000000000000000000000000000000000000000000000000000000";
 
      uint64[2] memory t;
      t[0] = 18446744073709551552;  // ffffffffffffffc0
      t[1] = 18446744073709551615;	// ffffffffffffffff

      bool f = true;

      uint32 rounds = 12;

      return F(h, m, t, f, rounds);
    }
```

And here is what is received in the precompile:
```
func (c *blake2F) Run(input []byte) ([]byte, error) {
	log.Info("F precompile", "len", len(input), "input", input, "input_hex", hex.EncodeToString(input))
```


> INFO [06-24|18:47:05.016] F precompile                             len=213 input="[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 213 106 9 230 39 243 188 201 9 187 103 174 132 132 202 167 59 60 110 243 114 254 148 184 43 165 79 245 58 95 29 54 242 81 14 82 127 173 230 130 209 155 5 104 140 43 62 108 31 31 131 217 171 251 65 189 107 91 224 205 25 19 126 33 121 39 132 0 52 14 107 5 197 117 37 146 197 44 79 18 18 146 234 252 81 204 1 169 151 180 174 209 52 9 41 143 173 13 153 204 200 247 100 83 217 179 102 30 92 77 50 93 119 81 20 125 177 116 137 4 109 22 130 213 14 239 164 161 218 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]" input_hex=00000000000000000000000000000000000000000000000000000000000000d56a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

My `args` are left-padded with `00000000000000000000000000000000000000000000000000000000000000d5` and for this reason they are truncated on the right side.

I checked and `abi.encodePacked` returns a correct value:

> 0x6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 0000000000000000000000ffffffffffffffc0ffffffffffffffff010000000c
            
Something wrong happens later and I have no idea yet what it is.

Reply:
I think it's that `bytes` is a dynamic array, 

> Arrays have a length member that contains their number of elements. The length of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.

So the `0x00...d5` is probably the length of the `bytes` array. @axic ?

Reply:
It is highly possible it is the lenght of the array but if so, it has to be added somewhere inside `staticcall`, maybe when we push args to client's memory because I do not see this prefix when I emit an event with the result of `abi.encodePacked`.

Reply:
You are using assembly above which gives a low level access to everything. One must be aware how types in memory are encoded. In `staticcall(..., args, ...)` the parameter `args` points to the memory area where that data resides, but in the case of `bytes memory` it is prefixed with the length.

In short, you need to use `staticcall(..., add(args, 32), ...)`.

Reply:
You were right, `add` helped. Thank you @axic and @holiman.